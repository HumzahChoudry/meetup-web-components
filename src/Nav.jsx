import React from 'react';

/**
 * Navigation component using SQ2 styles
 * @see {@link http://meetup.github.io/sassquatch2/}
 * @module Nav
 */
class Nav extends React.Component {
	render() {
		const {
			children,
			...other
		} = this.props;

		return (
			<nav
				role='navigation'
				{...other}>
				<ul className='row resetList'>
					{children}
				</ul>
			</nav>
		);
	}
}

export default Nav;

