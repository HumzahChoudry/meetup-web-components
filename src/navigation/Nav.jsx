import PropTypes from 'prop-types';
import React from 'react';
import cx from 'classnames';

import swarmLogo from '../../assets/svg/logo--mSwarm--2color.svg';
import scriptLogo from '../../assets/svg/logo--script.svg';
import withMatchMedia from '../utils/components/withMatchMedia';
import Flex from '../layout/Flex';
import FlexItem from '../layout/FlexItem';
import Icon from '../media/Icon';
import AvatarMember from '../media/AvatarMember';

import NavItem from './components/NavItem';
import ProfileDropdown from './components/profile/ProfileDropdown';
import NotificationsDropdown from './components/notifications/NotificationsDropdown';

const CLASS_UNAUTH_ITEM = 'navItem--unauthenticated';
const CLASS_AUTH_ITEM = 'navItem--authenticated';

export const DropdownLoader = ({ label }) => (
	<Section className="valignChildren--center align--center" aria-live="polite">
		<p className="visibility--a11yHide">{label}</p>
		<Icon shape="updates" size="l" />
	</Section>
);

/**
 * @param {Object} props component properties
 * @returns {React.element} Navigation Bar
 */
export class Nav extends React.Component {
	/**
	 * @return {React.element} the navbar component
	 */
	render() {
		const {
			media,
			self,
			logoAccessible,
			navItems,
			dropdownLoaderLabel,
			localeCode,
		} = this.props;
		const {
			login,
			signup,
			proDashboard,
			explore,
			messages,
			notifications,
			groups,
			profile,
		} = navItems;
		const isLoggedOut = self.status === 'prereg' || !self.name;

		const showScriptLogo = Boolean(media.isAtLargeUp || isLoggedOut);
		const showSwarmLogo = Boolean(
			media.isAtMediumUp && !media.isAtLargeUp && !isLoggedOut
		);
		const notificationContent = notifications.list ? (
			<NotificationsDropdown
				self={self}
				notifications={notifications.list}
				onMarkReadAction={notifications.notificationsDropdown.markRead}
				localeCode={localeCode}
				emptyContentLabel={notifications.notificationsDropdown.emptyContentLabel}
				notificationsTitleLabel={notifications.label}
				generateClassicUrl={
					notifications.notificationsDropdown.generateClassicUrl
				}
			/>
		) : (
			<DropdownLoader label={dropdownLoaderLabel} />
		);

		const profileContent = groups ? (
			<ProfileDropdown
				settings={profile.profileDropdown.settings}
				help={profile.profileDropdown.help}
				logout={profile.profileDropdown.logout}
				groupHome={profile.profileDropdown.groupHome}
				allGroupsLabel={profile.profileDropdown.allGroupsLabel}
				profile={profile}
				groups={groups.list}
			/>
		) : (
			<DropdownLoader label={dropdownLoaderLabel} />
		);

		let unauthItems = [
			{
				shrink: true,
				linkTo: login.link,
				label: login.label,
				className: `${CLASS_UNAUTH_ITEM} navItem--login`,
			},
			{
				shrink: true,
				onAction: () => {},
				label: signup.label,
				className: CLASS_UNAUTH_ITEM,
			},
		];

		let authItems = [
			self.isProMember && {
				shrink: true,
				linkTo: proDashboard.link,
				label: proDashboard.label,
				className: cx(
					'text--blue text--bold navItemLink--create display--none atMedium_display--block'
				),
			},
			{
				shrink: true,
				linkTo: explore.link,
				label: explore.label,
				className: cx(CLASS_AUTH_ITEM, { 'flush--left': !media.isAtMediumUp }),
				icon: <Icon shape="search" size="s" className="atMedium_display--none" />,
			},
			{
				shrink: true,
				linkTo: groups.link,
				label: groups.label,
				className: `atMedium_display--none ${CLASS_AUTH_ITEM}`,
				icon: <Icon shape="groups" size="s" className="display--block" />,
			},
			{
				shrink: true,
				linkTo: messages.link,
				label: messages.label,
				className: cx('navItem--messages', CLASS_AUTH_ITEM),
				icon: (
					<Icon
						shape="messages"
						size="s"
						className="display--block atMedium_display--none"
					/>
				),
				hasUpdates: messages.unreadMessages > 0,
			},
			{
				shrink: true,
				linkTo: media.isAtMediumUp ? false : notifications.link,
				label: notifications.label,
				className: cx('navItem--notifications', CLASS_AUTH_ITEM),
				icon: (
					<Icon
						shape="notifications"
						size="s"
						className="display--block atMedium_display--none"
					/>
				),
				onClickAction:
					media.isAtMediumUp && notifications.list.length == 0
						? notifications.getNotificationsQuery
						: false,
				dropdownContent: media.isAtMediumUp && notificationContent,
				hasUpdates: notifications.unreadNotifications > 0,
			},
			{
				shrink: true,
				linkTo: media.isAtMediumUp ? false : '/profile/',
				label: profile.Label,
				labelClassName: 'navItem-label display--block atMedium_display--none',
				className: cx(CLASS_AUTH_ITEM, 'profileDropdown', {
					'profileDropdown--hasGroups': Boolean(
						groups.list && groups.list.length
					),
				}),
				icon: (
					<Flex noGutters align="center">
						<FlexItem>
							<AvatarMember small member={self} />
						</FlexItem>
						<FlexItem
							shrink
							className="display--none atMedium_display--block"
						>
							<Icon
								shape="chevron-down"
								size="xxs"
								className="padding--left-half"
							/>
						</FlexItem>
					</Flex>
				),
				onClickAction:
					media.isAtMediumUp && groups.list.length == 0
						? profile.getSelfGroupsQuery
						: false,
				dropdownContent: media.isAtMediumUp && profileContent,
			},
		];

		unauthItems = unauthItems.map((item, i) => {
			return (
				<NavItem
					key={i}
					shrink={item.shrink}
					linkTo={item.linkTo && item.linkTo}
					label={item.label}
					className={item.className}
					onAction={item.onAction && item.onAction}
				/>
			);
		});

		authItems = authItems.map((item, i) => {
			return (
				item && (
					<NavItem
						key={i}
						shrink={item.shrink && item.shrink}
						linkTo={item.linkTo && item.linkTo}
						label={item.label && item.label}
						labelClassName={item.labelClassName && item.labelClassName}
						className={item.className && item.className}
						icon={item.icon && item.icon}
						onClickAction={item.onClickAction && item.onClickAction}
						dropdownContent={item.dropdownContent && item.dropdownContent}
						hasUpdates={item.hasUpdates && item.hasUpdates}
					/>
				)
			);
		});

		return (
			<nav
				aria-label="Header navigation"
				role="navigation"
				className="padding--all"
				id="globalNav"
			>
				<Flex
					align={media.isAtMediumUp ? 'center' : 'top'}
					justify="spaceBetween"
					className="span--100 navBar"
				>
					{showSwarmLogo && (
						<NavItem
							linkTo="meetup.com"
							className="logo logo--swarm flush--left"
							icon={
								<img src={swarmLogo} alt={logoAccessible} height="48px" />
							}
						/>
					)}

					{showScriptLogo && (
						<NavItem
							linkTo="meetup.com"
							className="logo logo--script flush--left"
							linkClassName="display--block"
							icon={
								<img
									src={scriptLogo}
									alt={logoAccessible}
									height="44px"
								/>
							}
						/>
					)}

					{isLoggedOut ? unauthItems : authItems}
				</Flex>
			</nav>
		);
	}
}

Nav.defaultProps = {
	logoAccessible: 'Meetup Logo',
	localeCode: 'en-US',
};
Nav.propTypes = {
	self: PropTypes.object,
	media: PropTypes.object,
	authItems: PropTypes.array,
	unauthItems: PropTypes.array,
	logoAccessible: PropTypes.string,
};

export default withMatchMedia(Nav);
