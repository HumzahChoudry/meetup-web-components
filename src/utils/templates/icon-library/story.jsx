export default (name, iconRefs, iconStories) => `
import React from 'react';
import Icon from './Icon';
import { Annotate, Inverted } from './utils/storyComponents';
import { storiesOf, linkTo } from '@kadira/storybook';

storiesOf('${name}', module)
	.add('default', () => (
		<div style={{
			width:'640px',
			marginTop: '450px',
			marginBottom: '50px'
		}}>
			<ul className='gridList gridList--has8'>
				${iconRefs}
			</ul>
		</div>
	))
	${iconStories}
	;
`;

