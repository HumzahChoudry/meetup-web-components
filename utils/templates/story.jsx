export default (name) => `
import React from 'react';
import ${name} from './${name}';
import { storiesOf } from '@kadira/storybook';

storiesOf('${name}', module)
	.add('default', () => <${name} />)
`;

