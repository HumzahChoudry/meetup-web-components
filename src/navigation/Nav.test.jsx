import React from 'react';
import { shallow } from 'enzyme';
import { MOCK_MEMBER } from 'meetup-web-mocks/lib/api';

import { Nav } from './Nav';
import { navItems } from './nav.story';

const MOCK_PROPS = {
	navItems,
	media: {
		isAtSmallUp: false,
		isAtMediumUp: false,
		isAtLargeUp: false,
	},
	self: { status: 'prereg' },
};

const wrapper = props => shallow(<Nav {...MOCK_PROPS} {...props} />);

describe('Nav', () => {
	it('should match the snapshot for unauthenticated small screens', () => {
		expect(wrapper()).toMatchSnapshot();
	});

	it('should match the snapshot for unauthenticated medium screens', () => {
		expect(
			wrapper({
				self: MOCK_MEMBER,
			})
		).toMatchSnapshot();
	});

	it('should match the snapshot for authenticated medium screens', () => {
		expect(
			wrapper({
				media: { isAtMediumUp: true },
				self: MOCK_MEMBER,
			})
		).toMatchSnapshot();
	});

	it('should match the snapshot for unauthenticated medium screens', () => {
		expect(
			wrapper({
				media: { isAtMediumUp: true },
			})
		).toMatchSnapshot();
	});

	it('should match the snapshot for unauthenticated large screens', () => {
		expect(wrapper({ media: { isAtLargeUp: true } })).toMatchSnapshot();
	});

	it('should match the snapshot for authenticated large screens', () => {
		expect(
			wrapper({
				media: { isAtLargeUp: true },
				self: MOCK_MEMBER,
			})
		).toMatchSnapshot();
	});

	it('should match the snapshot with logo photo', () => {
		expect(
			wrapper({
				self: MOCK_MEMBER,
				navItems: {
					...navItems,
					proDashboard: {
						mainAccount: {
							urlname: '/mason-mocks',
							name: 'Mason Mocks',
							group_photo: {
								thumb_link: 'https://placeimg.com/640/480/any',
							},
						},
					},
				},
			})
		).toMatchSnapshot();
	});

	it('should match the snapshot without a logo photo', () => {
		expect(
			wrapper({
				self: MOCK_MEMBER,
				navItems: {
					...navItems,
					mainAccount: {
						urlname: '/mason-mocks',
						name: 'Mason Mocks',
						group_photo: {},
					},
				},
			})
		).toMatchSnapshot();
	});
});
