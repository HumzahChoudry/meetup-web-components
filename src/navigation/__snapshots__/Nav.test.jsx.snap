// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav should match the snapshot for authenticated large screens 1`] = `
<Nav
  localeCode="en-US"
  media={Object {}}
  self={
    Object {
      "id": 1243,
      "name": "mocky mcmockerson",
      "photo": Object {
        "photo_link": "http://placekitten.com/g/400/400",
        "thumb_link": "http://placekitten.com/g/200/200",
      },
    }
  }
/>
`;

exports[`Nav should match the snapshot for authenticated medium screens 1`] = `
<Nav
  localeCode="en-US"
  media={Object {}}
  self={
    Object {
      "id": 1243,
      "name": "mocky mcmockerson",
      "photo": Object {
        "photo_link": "http://placekitten.com/g/400/400",
        "thumb_link": "http://placekitten.com/g/200/200",
      },
    }
  }
/>
`;

exports[`Nav should match the snapshot for unauthenticated large screens 1`] = `
<Nav
  localeCode="en-US"
  media={Object {}}
  self={
    Object {
      "status": "prereg",
    }
  }
/>
`;

exports[`Nav should match the snapshot for unauthenticated medium screens 1`] = `
<Nav
  localeCode="en-US"
  media={Object {}}
  self={
    Object {
      "id": 1243,
      "name": "mocky mcmockerson",
      "photo": Object {
        "photo_link": "http://placekitten.com/g/400/400",
        "thumb_link": "http://placekitten.com/g/200/200",
      },
    }
  }
/>
`;

exports[`Nav should match the snapshot for unauthenticated medium screens 2`] = `
<Nav
  localeCode="en-US"
  media={Object {}}
  self={
    Object {
      "status": "prereg",
    }
  }
/>
`;

exports[`Nav should match the snapshot for unauthenticated small screens 1`] = `
<Nav
  localeCode="en-US"
  media={Object {}}
  self={
    Object {
      "status": "prereg",
    }
  }
/>
`;
